### PR TITLE
NDRS-1567: Add get-account-info command

### PR DESCRIPTION
**Pull request by darthsiroftardis**
CHANGELOG:

- Added `get-account-info` command to the client
- Added `state_get_account_info` RPC method to the node which queries `GlobalState` for an `Account`

### CODE REVIEW

**Review by nilz3ro — APPROVED** [review]
Looks good to me, I'm still new to this part of the codebase though.

**Review by EdHastingsCasperLabs — APPROVED** [review]

**Review by Fraser999 — COMMENTED** [review]
Just nitpicks from me.  Nice work.

File: client/lib/lib.rs [hunk 1]
```diff
@@ -464,6 +464,31 @@ pub fn get_auction_info(
...
+///   will not be printed to `stdout`.
+/// * `public_key` the public key associated with the `Account`
+/// * `maybe_block_id` must be a hex-encoded, 32-byte hash digest or a `u64` representing the
+///   `Block` height or empty. If empty, the latest `Block` will be retrieved.
+pub fn get_account_info(
+    maybe_rpc_id: &str,
+    node_address: &str,
+    verbosity_level: u64,
+    public_key: &str,
+    block_id: &str,
```
[the remark this answers is not in the document]
Comment (darthsiroftardis) [reply client/lib/lib.rs]: Renamed as suggested

File: client/lib/rpc.rs [hunk 2]
```diff
@@ -243,6 +243,20 @@ impl RpcCall {
         Ok(response)
     }
 
+    pub(crate) fn get_account_info(self, public_key: &str, block_id: &str) -> Result<JsonRpc> {
```
[the remark this answers is not in the document]
Comment (darthsiroftardis) [reply client/lib/rpc.rs]: Renamed as suggested

File: client/src/get_account_info.rs [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+enum DisplayOrder {
+    Verbose,
+    NodeAddress,
+    RpcId,
+    Key,
+    BlockIdentifier,
+}
+
+/// Handles providing the arg for and retrieval of the public key.
+mod public_key {
```
[the remark this answers is not in the document]
Comment (darthsiroftardis) [reply client/src/get_account_info.rs]: Moved to `common` and `arg()` takes display order as a parameter

File: client/src/main.rs [hunk 4]
```diff
@@ -54,6 +55,7 @@ enum DisplayOrder {
     GenerateCompletion,
     GetRpcs,
     AccountAddress,
+    GetAccountInfo,
```
[the remark this answers is not in the document]
Comment (darthsiroftardis) [reply client/src/main.rs]: Moved

File: client/src/main.rs [hunk 5]
```diff
@@ -87,6 +89,7 @@ fn cli<'a, 'b>() -> App<'a, 'b> {
         ))
         .subcommand(ListRpcs::build(DisplayOrder::GetRpcs as usize))
         .subcommand(AccountAddress::build(DisplayOrder::AccountAddress as usize))
+        .subcommand(GetAccountInfo::build(DisplayOrder::GetAccountInfo as usize))
```
[the remark this answers is not in the document]
Comment (darthsiroftardis) [reply client/src/main.rs]: Moved

File: client/src/main.rs [hunk 6]
```diff
@@ -114,6 +117,7 @@ async fn main() {
         (GenerateCompletion::NAME, Some(matches)) => (GenerateCompletion::run(matches), matches),
         (ListRpcs::NAME, Some(matches)) => (ListRpcs::run(matches), matches),
         (AccountAddress::NAME, Some(matches)) => (AccountAddress::run(matches), matches),
+        (GetAccountInfo::NAME, Some(matches)) => (GetAccountInfo::run(matches), matches),
```
[the remark this answers is not in the document]
Comment (darthsiroftardis) [reply client/src/main.rs]: Moved

File: node/src/components/rpc_server/rpcs/state.rs [hunk 7]
```diff
@@ -412,3 +427,132 @@ impl RpcWithOptionalParamsExt for GetAuctionInfo {
...
+}
+
+/// Result for "state_get_account_info" RPC response.
+#[derive(Serialize, Deserialize, Debug, JsonSchema)]
+#[serde(deny_unknown_fields)]
+pub struct GetAccountInfoResult {
+    /// The RPC API version.
+    #[schemars(with = "String")]
+    pub api_version: ProtocolVersion,
+    /// The Account
```
[the remark this answers is not in the document]
Comment (darthsiroftardis) [reply node/src/components/rpc_server/rpcs/state.rs]: Fixed typo

File: node/src/components/rpc_server/rpcs/state.rs [hunk 8]
```diff
@@ -412,3 +427,132 @@ impl RpcWithOptionalParamsExt for GetAuctionInfo {
...
+    pub account: Account,
+    /// THe merkle proof.
```
[the remark this answers is not in the document]
Comment (darthsiroftardis) [reply node/src/components/rpc_server/rpcs/state.rs]: Fixed

File: node/src/components/rpc_server/rpcs/state.rs [hunk 9]
```diff
@@ -412,3 +427,132 @@ impl RpcWithOptionalParamsExt for GetAuctionInfo {
...
+                            responder,
+                        },
+                        QueueKind::Api,
+                    )
+                    .await;
+
+                match maybe_block {
+                    None => {
+                        let error_msg =
+                            "get-account-info failed to get specified block".to_string();
```
[the remark this answers is not in the document]
Comment (darthsiroftardis) [reply node/src/components/rpc_server/rpcs/state.rs]: Used pattern from `GetAuctionInfo`

File: node/src/components/rpc_server/rpcs/state.rs [hunk 10]
```diff
@@ -412,3 +427,132 @@ impl RpcWithOptionalParamsExt for GetAuctionInfo {
...
+                        return Ok(response_builder.error(warp_json_rpc::Error::custom(
+                            ErrorCode::NoSuchBlock as i64,
+                            error_msg,
+                        ))?);
+                    }
+                    Some((block, _)) => block,
+                }
+            };
+
+            // retrieve the global state hash of the block.
```
[the remark this answers is not in the document]
Comment (darthsiroftardis) [reply node/src/components/rpc_server/rpcs/state.rs]: Removed comment

File: node/src/components/rpc_server/rpcs/state.rs [hunk 11]
```diff
@@ -412,3 +427,132 @@ impl RpcWithOptionalParamsExt for GetAuctionInfo {
...
+                    info!("{}", error_msg);
+                    return Ok(response_builder
+                        .error(warp_json_rpc::Error::custom(error_code as i64, error_msg))?);
+                }
+            };
+
+            let account = if let StoredValue::Account(account) = stored_value {
+                account
+            } else {
+                let error_msg = "state_get_account could not retrieve an Account".to_string();
```
[the remark this answers is not in the document]
Comment (darthsiroftardis) [reply node/src/components/rpc_server/rpcs/state.rs]: Maintained consistency and changed error variant

**Review by Fraser999 — APPROVED** [review]

**Pull request merged**